PBS server reports expired subjobs to scheduler

arungrover opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Bug/feature Description
While fixing the issue in PR #791 I added a new log message in scheduler which is logged if scheduler find a subjob in invalid state (!RUNNING). Along with this server was also changed to not report non-running subjobs to scheduler. But that part of the fix is not working and as a result, scheduler logs a bunch of messages about subjobs being invalid every cycle when job history is enabled.

#### Affected Platform(s)
All

#### Cause / Analysis / Design
In server the check to bail out if the subjob was a not running subjob was placed at a wrong place.

#### Solution Description
Moved the check to the right place.

#### Testing logs/output
[test_without_fix.txt](https://github.com/PBSPro/pbspro/files/2551199/test_without_fix.txt)
[test_out.txt](https://github.com/PBSPro/pbspro/files/2551200/test_out.txt)
[test_smoketest.txt](https://github.com/PBSPro/pbspro/files/2551201/test_smoketest.txt)


#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

agrawalravi90 commented on test/tests/functional/pbs_array_job.py:
```diff
@@ -0,0 +1,72 @@
+from tests.functional import *
+
+
+class TestArrayJobs(TestFunctional):
```
There already is a file called "pbs_job_array.py", please add this test to that test suite.

arungrover replied:
Thanks for pointing me to that file, I'll move this test

agrawalravi90 commented on test/tests/functional/pbs_array_job.py:
```diff
@@ -0,0 +1,72 @@
+    """
+    def test_expired_subjobs_not_reported(self):
+        """
+        Test if a subjob is finished ans moves to expired state,
```
typo and* instead of ans

arungrover replied:
oops

agrawalravi90 commented on test/tests/functional/pbs_array_job.py:
```diff
@@ -0,0 +1,72 @@
+        res_req = {'Resource_List.select': '1:ncpus=1',
+                   'array_indices_submitted': '1-6'}
+        j1 = Job(TEST_USER, attrs=res_req)
+        j1.set_sleep_time(2)
```
2 seconds seems to small to me for slower machines, how about turning scheduling off after submitting the job, doing expect to see if it expired, then triggering the next cycle to search for the log message?

anamikau replied:
@agrawalravi90 you are right in general. but here since we are not looking for job state in R but only X it is safe.

arungrover replied:
My intent is to have some subjobs in 'X' state and some in 'Q' or 'R'. Also, to have the test not take a long time I reduced the time interval to 2 seconds. I'd like to retain as is.

anamikau replied:
@arungrover if that is your intent then I suggest submit a job array requesting 100 subjobs or so.

arungrover replied:
Ok, I can increase the number of subjobs here

anamikau commented on test/tests/functional/pbs_array_job.py:
```diff
@@ -0,0 +1,72 @@
+        """
+
+        a = {'resources_available.ncpus': 1}
+        self.server.manager(MGR_CMD_SET, NODE, a, self.mom.shortname)
```
why do you need to set ncpus=1? looking at your test it doesn't matter if you have 1 or more.

arungrover replied:
The reason is that I want some of the subjobs to be either in running state or queued state. This is why I made the available ncpus as 1 because then only one job can run at a time.

anamikau replied:
okay but then this configuration would be hard to run on a CRAY or cpuset systems. why don't you submit an exclusive job so that PBS cannot run more than one job one a node?

arungrover replied:
Well, you are right, but even if I take the node exclusively it might still not work because those systems come with loads of vnodes prebuilt. What I can probably do is that I can submit a job requesting mom name as the node where the job needs to run. This way it will only run on login node.

arungrover commented on test/tests/functional/pbs_array_job.py:
```diff
@@ -0,0 +1,72 @@
+class TestArrayJobs(TestFunctional):
+
+    """
+    This test suite tests if PBS's array job functionality
```
Ah.. I missed adding 'working properly' at the end of the sentence. removing 'if' will also work in this case. I'll change it;

arungrover commented:
Thanks @agrawalravi90 and @anamikau  for your prompt review. I have addressed your comments and here is the test log - 
[test_out.txt](https://github.com/PBSPro/pbspro/files/2559527/test_out.txt)

anamikau commented on test/tests/functional/pbs_job_array.py:
```diff
@@ -610,3 +610,32 @@ def test_job_id_duplicate_server_restart(self):
+        """
+
+        a = {'resources_available.ncpus': 1}
+        self.server.manager(MGR_CMD_SET, NODE, a, self.mom.shortname)
```
I thought we decided to remove this and rather submit the jobs with exclusive flag with host=mom node to make it more generic.

arungrover replied:
Thanks for pointing that out Anamika, I've pushed that change in

agrawalravi90 commented:
Looks good to me as well, thanks.

mike0042 commented:
Looks like the Appveyor release build timed out. Please squash and rebase, and I suspect it will work.

mike0042 commented:
This time the debug build failed @arungrover. Please re-sign your commit to restart the CI checks. I also notice that the Codacy check isn't listed, but I saw that it passed previously.

arungrover commented:
@mike0042 I restarted it, I don't know why debug build is failing.

arungrover commented:
Thanks @mike0042
